File: tests/WeavePort.ConcurrentSdkTests/fixtures/streams.py
import asyncio
import os
import sys
from pathlib import Path
sdk = sys.argv[sys.argv.index('--sdk-path') + 1] if '--sdk-path' in sys.argv else str(Path(__file__).resolve().parents[3] / 'sdks/python')
sys.path.insert(0, sdk)
from weaveport_sdk import PluginApplication
app = PluginApplication('1')
@app.function('echo')
async def echo(value, context):
    await asyncio.sleep(value.get('delay', 0) / 1000)
    return dict(pid=os.getpid(), tenant=context.tenant)
@app.stream('slow')
async def slow(value, context):
    await asyncio.sleep(value.get('firstDelay', 0) / 1000)
    yield dict(pid=os.getpid(), item=1)
    await asyncio.sleep(value.get('nextDelay', 300) / 1000)
    yield dict(pid=os.getpid(), item=2)
@app.function('identity')
async def identity(value, context):
    return dict(tenant=context.tenant)
@app.function('stats')
async def stats(value, context):
    return dict(pid=os.getpid())
@app.stream('live')
async def live(value, context):
    yield 1
    await asyncio.sleep(.3)
    await context.call_host('echo', value)
    yield 2
class PatternSource:
    def __init__(self, count):
        self.remaining = count
    def read(self, size):
        count = min(self.remaining, size)
        self.remaining -= count
        return b'Z' * count
    def close(self):
        pass
@app.source('large')
async def large(value, context):
    return PatternSource(value['bytes'])
app.run()

File: tests/WeavePort.ConcurrentSdkTests/fixtures/streams.mjs
import { pathToFileURL } from 'node:url';
const sdkIndex = process.argv.indexOf('--sdk-path');
const { PluginApplication } = await import(sdkIndex >= 0 ? pathToFileURL(process.argv[sdkIndex + 1]).href : '../../../sdks/typescript/dist/index.js');
const app = new PluginApplication('1');
app.function('echo', async (value, context) => {
    await new Promise(resolve => setTimeout(resolve, value.delay ?? 0));
    return {pid: process.pid, tenant: context.tenant};
});
app.stream('slow', async function* (value) {
    await new Promise(resolve => setTimeout(resolve, value.firstDelay ?? 0));
    yield {pid: process.pid, item: 1};
    await new Promise(resolve => setTimeout(resolve, value.nextDelay ?? 300));
    yield {pid: process.pid, item: 2};
});
app.function('identity', async (value, context) => ({tenant: context.tenant}));
app.function('stats', async () => ({pid: process.pid}));
app.stream('live', async function* (value, context) {
    yield 1;
    await new Promise(resolve => setTimeout(resolve, 300));
    await context.callHost('echo', value);
    yield 2;
});
app.stream('immediateCallback', async function* (value, context) {
    yield 1;
    await context.callHost('echo', value);
    yield 2;
});
app.source('large', async value => {
    let remaining = value.bytes;
    return {
        read(size) { const count = Math.min(remaining, size); remaining -= count; return new Uint8Array(count).fill(0x5a); },
        close() { }
    };
});
await app.run();
